File: sdk/python/src/attestplane/settlement_verifier.py
# SPDX-FileCopyrightText: 2026 The Attestplane Authors
# SPDX-License-Identifier: Apache-2.0
"""Settlement-precondition verifier predicate — read-only walker, NEVER settles.

ADR-0009 § B.3 + P2.3. Given a :class:`SettlementPreconditionClaim`
asserting "lease X was consumed before settlement Y was requested in
this chain", this module's :func:`check_settlement_precondition`
function walks the supplied chain and confirms (or rejects) the claim.

**Hard constraint** (per ADR-0009 § B.3 + invariant 7 + REDLINE C.3
+ C.8): this module NEVER executes settlement, NEVER allocates budget,
NEVER mutates state. It only walks the chain looking for an ordered
pair of observations (lease consumed, then settlement requested) and
returns a structured result.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any


@dataclass(frozen=True, slots=True)
class SettlementPreconditionClaim:
    """A claim that the chain satisfies verify-before-settle.

    Mirrors schemas/v1/settlement_precondition_claim.schema.json.
    The verifier walks for:

    - A ``lease_lifecycle_event`` with payload.lifecycle == "consumed"
      and payload.lease_id_hash == self.lease_id_hash.
    - A ``settlement_event`` with payload.settlement_run_id ==
      self.settlement_run_id.
    - The lease-consumed event MUST have a strictly lower seq than the
      settlement_event.
    """

    claim_kind: str  # always "settlement_precondition" in v1
    lease_id_hash: str
    settlement_run_id: str
    expected_settlement_amount_hash: str | None = None
    claim_observed_at: str | None = None


@dataclass(frozen=True, slots=True)
class SettlementVerificationResult:
    """Outcome of :func:`check_settlement_precondition`.

    - ``ok=True``: both observations present and correctly ordered.
    - ``ok=False`` with reason:
      * "lease_consumed_not_observed"
      * "settlement_event_not_observed"
      * "settlement_precedes_lease_consumed" (ordering violation)
      * "claim_kind_unsupported"
      * "amount_hash_mismatch" (if claim provided expected_settlement_amount_hash
        and the settlement_event's amount_hash field differs)
    """

    ok: bool
    reason: str | None
    lease_consumed_seq: int | None
    settlement_event_seq: int | None


def check_settlement_precondition(
    chain_events: list[dict[str, Any]],
    claim: SettlementPreconditionClaim,
    *,
    verification_time: datetime | None = None,
) -> SettlementVerificationResult:
    """Verify the claim against the supplied chain. Read-only. Never settles.

    :param chain_events: list of event dicts. Each MUST contain
        ``seq``, ``event_type``, ``payload``. (This matches the
        JSONL backend's serialised form and ProofBundle's events array.)
    :param claim: the claim to verify.
    :param verification_time: reserved for future window checks; v1
        ignores. Must be UTC-aware if supplied (defensive check).
    """
    if claim.claim_kind != "settlement_precondition":
        return SettlementVerificationResult(
            ok=False,
            reason=f"claim_kind_unsupported: {claim.claim_kind!r}",
            lease_consumed_seq=None,
            settlement_event_seq=None,
        )
    if verification_time is not None and verification_time.tzinfo is None:
        return SettlementVerificationResult(
            ok=False,
            reason="verification_time must be UTC-aware",
            lease_consumed_seq=None,
            settlement_event_seq=None,
        )
    if not isinstance(chain_events, list):
        return SettlementVerificationResult(
            ok=False,
            reason=f"chain_events must be list, got {type(chain_events).__name__}",
            lease_consumed_seq=None,
            settlement_event_seq=None,
        )

    lease_consumed_seq: int | None = None
    settlement_event_seq: int | None = None
    settlement_amount_hash: str | None = None

    for ev in chain_events:
        if not isinstance(ev, dict):
            continue
        seq = ev.get("seq")
        if not isinstance(seq, int):
            continue
        event_type = ev.get("event_type")
        payload = ev.get("payload")
        if not isinstance(payload, dict):
            continue

        if event_type == "lease_lifecycle_event":
            if (
                payload.get("lifecycle") == "consumed"
                and payload.get("lease_id_hash") == claim.lease_id_hash
            ):
                if lease_consumed_seq is None or seq < lease_consumed_seq:
                    lease_consumed_seq = seq
        elif event_type == "settlement_event":
            if payload.get("settlement_run_id") == claim.settlement_run_id:
                if settlement_event_seq is None or seq < settlement_event_seq:
                    settlement_event_seq = seq
                    settlement_amount_hash = payload.get("amount_hash")

    if lease_consumed_seq is None:
        return SettlementVerificationResult(
            ok=False,
            reason="lease_consumed_not_observed",
            lease_consumed_seq=None,
            settlement_event_seq=settlement_event_seq,
        )
    if settlement_event_seq is None:
        return SettlementVerificationResult(
            ok=False,
            reason="settlement_event_not_observed",
            lease_consumed_seq=lease_consumed_seq,
            settlement_event_seq=None,
        )
    if lease_consumed_seq >= settlement_event_seq:
        return SettlementVerificationResult(
            ok=False,
            reason=(
                f"settlement_precedes_lease_consumed: "
                f"lease consumed at seq={lease_consumed_seq}, "
                f"settlement requested at seq={settlement_event_seq}"
            ),
            lease_consumed_seq=lease_consumed_seq,
            settlement_event_seq=settlement_event_seq,
        )
    if (
        claim.expected_settlement_amount_hash is not None
        and settlement_amount_hash is not None
        and settlement_amount_hash != claim.expected_settlement_amount_hash
    ):
        return SettlementVerificationResult(
            ok=False,
            reason=(
                f"amount_hash_mismatch: claim expected "
                f"{claim.expected_settlement_amount_hash}, settlement "
                f"event reports {settlement_amount_hash}"
            ),
            lease_consumed_seq=lease_consumed_seq,
            settlement_event_seq=settlement_event_seq,
        )

    return SettlementVerificationResult(
        ok=True,
        reason=None,
        lease_consumed_seq=lease_consumed_seq,
        settlement_event_seq=settlement_event_seq,
    )


__all__ = [
    "SettlementPreconditionClaim",
    "SettlementVerificationResult",
    "check_settlement_precondition",
]

File: sdk/typescript/src/settlement_verifier.ts
// SPDX-FileCopyrightText: 2026 The Attestplane Authors
// SPDX-License-Identifier: Apache-2.0
/**
 * Settlement-precondition verifier predicate — read-only walker, NEVER settles.
 *
 * ADR-0009 § B.3 + P2.3. Walks the chain for an ordered pair of
 * observations: lease consumed → settlement requested (same lease &
 * settlement IDs). Returns yes/no/reason. NEVER executes settlement.
 *
 * **Hard constraint** (per ADR-0009 § B.3 + invariant 7 + REDLINE C.3
 * + C.8): no settlement execution, no budget allocation, no state
 * mutation. Pure function.
 */

export interface SettlementPreconditionClaim {
  readonly claim_kind: 'settlement_precondition' | string;
  readonly lease_id_hash: string;
  readonly settlement_run_id: string;
  readonly expected_settlement_amount_hash?: string;
  readonly claim_observed_at?: string;
}

export interface SettlementVerificationResult {
  readonly ok: boolean;
  readonly reason: string | null;
  readonly lease_consumed_seq: number | null;
  readonly settlement_event_seq: number | null;
}

/** Minimal chain-event shape this verifier needs. */
export interface ChainEventForSettlement {
  readonly seq: number;
  readonly event_type: string;
  readonly payload: Record<string, unknown>;
}

const HEX64 = /^[0-9a-f]{64}$/;

export function checkSettlementPrecondition(
  chainEvents: readonly ChainEventForSettlement[],
  claim: SettlementPreconditionClaim,
  options: { readonly verification_time?: Date } = {},
): SettlementVerificationResult {
  if (claim.claim_kind !== 'settlement_precondition') {
    return {
      ok: false,
      reason: `claim_kind_unsupported: ${JSON.stringify(claim.claim_kind)}`,
      lease_consumed_seq: null,
      settlement_event_seq: null,
    };
  }
  if (
    options.verification_time !== undefined &&
    Number.isNaN(options.verification_time.getTime())
  ) {
    return {
      ok: false,
      reason: 'verification_time must be a valid Date',
      lease_consumed_seq: null,
      settlement_event_seq: null,
    };
  }
  if (!Array.isArray(chainEvents)) {
    return {
      ok: false,
      reason: `chain_events must be array, got ${typeof chainEvents}`,
      lease_consumed_seq: null,
      settlement_event_seq: null,
    };
  }

  let leaseConsumedSeq: number | null = null;
  let settlementEventSeq: number | null = null;
  let settlementAmountHash: string | null = null;

  for (const ev of chainEvents) {
    if (ev === null || typeof ev !== 'object') continue;
    const seq = ev.seq;
    if (typeof seq !== 'number' || !Number.isInteger(seq)) continue;
    const payload = ev.payload;
    if (payload === null || typeof payload !== 'object') continue;

    if (ev.event_type === 'lease_lifecycle_event') {
      if (
        (payload as Record<string, unknown>).lifecycle === 'consumed' &&
        (payload as Record<string, unknown>).lease_id_hash === claim.lease_id_hash
      ) {
        if (leaseConsumedSeq === null || seq < leaseConsumedSeq) {
          leaseConsumedSeq = seq;
        }
      }
    } else if (ev.event_type === 'settlement_event') {
      if ((payload as Record<string, unknown>).settlement_run_id === claim.settlement_run_id) {
        if (settlementEventSeq === null || seq < settlementEventSeq) {
          settlementEventSeq = seq;
          settlementAmountHash =
            (typeof (payload as Record<string, unknown>).amount_hash === 'string'
              ? ((payload as Record<string, unknown>).amount_hash as string)
              : null) ?? null;
        }
      }
    }
  }

  if (leaseConsumedSeq === null) {
    return {
      ok: false,
      reason: 'lease_consumed_not_observed',
      lease_consumed_seq: null,
      settlement_event_seq: settlementEventSeq,
    };
  }
  if (settlementEventSeq === null) {
    return {
      ok: false,
      reason: 'settlement_event_not_observed',
      lease_consumed_seq: leaseConsumedSeq,
      settlement_event_seq: null,
    };
  }
  if (leaseConsumedSeq >= settlementEventSeq) {
    return {
      ok: false,
      reason: `settlement_precedes_lease_consumed: lease consumed at seq=${leaseConsumedSeq}, settlement requested at seq=${settlementEventSeq}`,
      lease_consumed_seq: leaseConsumedSeq,
      settlement_event_seq: settlementEventSeq,
    };
  }
  if (
    claim.expected_settlement_amount_hash !== undefined &&
    !HEX64.test(claim.expected_settlement_amount_hash)
  ) {
    return {
      ok: false,
      reason: 'expected_settlement_amount_hash_malformed',
      lease_consumed_seq: leaseConsumedSeq,
      settlement_event_seq: settlementEventSeq,
    };
  }
  if (claim.expected_settlement_amount_hash !== undefined) {
    if (settlementAmountHash === null) {
      return {
        ok: false,
        reason: 'amount_hash_missing',
        lease_consumed_seq: leaseConsumedSeq,
        settlement_event_seq: settlementEventSeq,
      };
    }
    if (!HEX64.test(settlementAmountHash)) {
      return {
        ok: false,
        reason: `amount_hash_malformed: settlement event reports ${JSON.stringify(settlementAmountHash)}`,
        lease_consumed_seq: leaseConsumedSeq,
        settlement_event_seq: settlementEventSeq,
      };
    }
    if (settlementAmountHash !== claim.expected_settlement_amount_hash) {
      return {
        ok: false,
        reason: `amount_hash_mismatch: claim expected ${claim.expected_settlement_amount_hash}, settlement event reports ${settlementAmountHash}`,
        lease_consumed_seq: leaseConsumedSeq,
        settlement_event_seq: settlementEventSeq,
      };
    }
  }

  return {
    ok: true,
    reason: null,
    lease_consumed_seq: leaseConsumedSeq,
    settlement_event_seq: settlementEventSeq,
  };
}
